### ISSUE TITLE
About WSL taskbar icons and mintty-2.9

### ISSUE THREAD

**Issue by angelog0** [mintty/mintty#784]
To start WSL, I use wslbridge (0.2.4) with mintty from my MSYS2 installation (maybe they are 2 years). I did a Windows link (WSLtty.lnk) on desktop and its target is

`C:\msys64\usr\bin\mintty.exe --position 20,450 --WSL=Ubuntu -~`

then I added it to the toolbar. In this way clicking on the toolbar icon of that link starts WSL and that icon is _underlined_ as usual.

Today there was the update from mintty-2.8.5 to mintty-2.9.0 and now when I click on that icon, another icon (identical) shows up. I have attached the images ([wslb-mintty.tar.gz](https://github.com/mintty/mintty/files/2253741/wslb-mintty.tar.gz)):

`wslb-mint-2.8.png`   what happens with mintty-2.8.5 (the xpected)

`wslb-mint-2.9.png`   what happens with mintty-2.9.0 (`NOT` expected)

Notice that with mintty-2.9 clicking with mouse-R on the 1st icon (that NOT underlined) shows

```
WSLtty                                            [with its WSL Ubuntu icon]
Remove from the toolbar
```

Instead, the same on 2nd icon (that which IS underlined) shows

```
MSYS2 Terminal                             [with MSYS2 icon!]
Add to the toolbar
Close the window
```

Definitely something is wrong: as I wrote, they are about two year I use wslbridge+mintty and never got these issues.

Maybe you have some ideas about this.

 Angelo

**Comment by mintty** [comment]
Please update wslbridge to the latest repository version (which is unfortunately not a release).

**Comment by Biswa96** [comment]
Does the `/bin` folder contains `cygwin-console-helper.exe` file with `cygwin1.dll`?

**Comment by angelog0** [comment]
@Biswa96, I wrote about MSYS2 not Cygwin. In any case I have found the first file:

```
$ ls  /bin/cyg*
/bin/cygcheck.exe  /bin/cygpath.exe  /bin/cygwin-console-helper.exe
```
So what else do you suggest? Thanks.

**Comment by angelog0** [comment]
@mintty, just for completeness, the issue shows up only for mintty-2.9. Reverting to 2.8.5 fixes it.

To summarize... 

I have MSYS2 installed and `WSLtty.lnk` (1) on the Desktop. I added that link to the TaskBar (TB)

Now when I click it on the TB another icons shows up in the TB. Clicking with mouse-R (right) button on the first icon (that I added) shows the image  `mouse-R-icon01.png` (1), which is expected. If I click with mouse-R the second icon the the image is `mouse-R-icon02.png` (1) which, in my opinion, is totally wrong.

In short, with 2.9 these [tips ](https://github.com/mintty/mintty/wiki/Tips#manual-setup-of-wsl-terminal) are useless.

--
(1)
[mintty-2.9-WSL.tar.gz](https://github.com/mintty/mintty/files/2255357/mintty-2.9-WSL.tar.gz)

**Comment by mintty** [comment]
That only confirms my previous comment. Somehow the contradictuous cases of Windows AppID configuration will have to be sorted out. I don't know if it's possible to achieve that fully automatically.

**Comment by angelog0** [comment]
@mintty, what you suggest seems to work as expected.

I have created a Desktop link (WSLTest.lnk) with target:

`C:\msys64\usr\bin\mintty.exe --position 20,450 --wsl /bin/wslbridge -t -C~ /bin/bash`

then I added it to the Task Bar. Clicking on it does not produce a second icon in TB and mouse-R on it displays the right item

  ```
WSLTest
  Remove from the TB
  Close the window
```

The reference to MSYS Terminal has gone :-)

**Comment by angelog0** [comment]
I wonder if the following is related to this issue..

After I Install `MSYS2` for the first time (last time was last year, Jul 2017) I add some personal links to `Start Menu`, under the `MSYS2` folder:

 ```
$ ls /c/ProgramData/Microsoft/Windows/Start\ Menu/Programs/MSYS2/
autorebase.lnk  MSYS2 MINGW32 Shell.lnk  MSYS2 MINGW64 Shell.lnk  MSYS2 MSYS Shell.lnk
```

To do this I had just created a tar ball installed (administrator: the prompt start with `#` from `getent` package) as:

```
# tar -xvof MSYS2-Start_Menu_Programs.tar.xz -C /c/ProgramData/Microsoft/Windows/Start\ Menu/Programs/
MSYS2/
MSYS2/autorebase.lnk
MSYS2/MSYS2 MINGW32 Shell.lnk
MSYS2/MSYS2 MINGW64 Shell.lnk
MSYS2/MSYS2 MSYS Shell.lnk
```

The folder `.../Programs/MSYS2` immediately showed up in `Start Menu` under `M` letter. Then I added those link to Task Bar (TB) and in Start. The target of `MSYS2 MSYS Shell.lnk` is for example:

`C:\msys64\msys2.exe`

where `msys2.exe` is from `msys2-launcher-git` package.

Now I want to rename `MSYS2 MSYS Shell.lnk` in `MSYS Terminal.lnk` and to test this I tried (as adm.)

```
# mkdir -p /c/ProgramData/Microsoft/Windows/Start\ Menu/Programs/test-MSYS2

# cp /c/ProgramData/Microsoft/Windows/Start\ Menu/Programs/MSYS2/MSYS2\ MSYS\ Shell.lnk /c/ProgramData/Microsoft/Windows/Start\ Menu/Programs/test-MSYS2/MSYS\ Terminal.lnk

```

The fact is that now the new folder `.../Programs/test-MSYS2` **DOES NOT** show up in `Start` (under `M` there is still the old `MSYS2`) and when I reach the new link `.../Programs/test-MSYS2/MSYS\ Terminal.lnk` (for example with `Open path to file` for another link in `Start`) it still refers to `MSYS2 MSYS Shell.lnk` even removing the others from TB, rebooting etc. 

`mouse-R` on it shows

MSYS2 MSYS Shell
Add to the toolbar
Close the window

(I would expect `MSYS Terminal`). I tries also to change the icon for `MSYS Terminal.lnk` but it uses the old.

All this is very confusing and frustrating, I doubt that reinstalling from scratch all works as before.

Maybe there is also some Windows (10) responsibility...

**Comment by mintty** [comment]
> mkdir -p /c/ProgramData/Microsoft/Windows/Start\ Menu/Programs/test-MSYS2

This should show up immediately. Note you're doing this in the "All Users" domain, maybe you can try in your personal Start Menu folder. Anyway, this is a plain Windows mechanism, not a mintty issue.

>  it still refers to ...

Of course it does. You simply copied a .lnk file, including all its information. To change it, open its Properties and change the Target, Icon, whatever...

**Comment by angelog0** [comment]
@mintty wrote:

> This should show up immediately. Note you're doing this in the "All Users" domain

Yes, in an _ideal world_! It seems that is broken here...

Thanks.

**Comment by mintty** [comment]
Released 2.9.1. New option value `AppID=@`.

**Comment by angelog0** [comment]
> Released 2.9.1. New option value `AppID=@`.

How should we use this option to create a good Desktop link to start WSL? Maybe the `Tips` section in the Wiki has to be updated.

Thanks.

**Comment by mintty** [comment]
Planning to improve the situation for the next release in two ways:
* Alt+F2 in a window started from a shortcut could invoke the shortcut rather than just the binary (.exe), thus inheriting the same AppID from the shortcut as the running instance.
* Add placeholders to option AppID to be able to use it but distinguish various deployments, invocations etc as preferred (somewhat covered by value AppID=@ already but more flexible).

**Comment by mintty** [comment]
Released 2.9.4.

**Comment by mintty** [comment]
Released 2.9.5.

**Comment by mintty** [comment]
I've uploaded a patch that derives the AppID from the desktop shortcut if mintty is started from a shortcut (desktop, taskbar or start menu).
Currently it takes precedence over an explicitly defined AppID option.
Please test.

**Comment by angelog0** [comment]
@mintty wrote:
> Please test.

Current master seems to work as expected.

**Comment by mintty** [comment]
Thanks; actually, to get this clear: does it work without setting the `AppID` option? Please confirm.

**Comment by angelog0** [comment]
Oops... I replayed to #801..

I start WSL with a link on the TB and whose target is:

```
C:\msys64\usr\bin\mintty.exe -c C:\msys64\home\user\.minttyrc -t "WSL Terminal" --position 20,450 --WSL=Ubuntu -~
```
and 

```
$ cat .minttyrc
BoldAsFont=-1
CursorType=block
Transparency=high
OpaqueWhenFocused=yes
Font=Lucida Console
FontHeight=12
Columns=112
Rows=30
X=70
Y=30
```

**Comment by mintty** [comment]
Released 2.9.6.
